presto-main/src/main/java/com/facebook/presto/operator/HashBuilderOperator.java

```diff
@@ -341,10 +385,15 @@ private void updateIndex(Page page)
             localRevocableMemoryContext.setBytes(index.getEstimatedSize().toBytes());
         }
         else {
+            long currentMemoryUsage = localUserMemoryContext.getBytes();
             if (!localUserMemoryContext.trySetBytes(index.getEstimatedSize().toBytes())) {
                 index.compact();
                 localUserMemoryContext.setBytes(index.getEstimatedSize().toBytes());
             }
+            if (enforceBroadcastMemoryLimit) {
```

Comment:
we actually want to do this as part of trySetBytes so that we compact the index if needed.  I would recommend having an overload of trySetbytes and setBytes (and then tryUpdateBytes and updateBytes) that takes in a boolean of whether to enforce the broadcast limit. 

Additionally rather than calling the full chain of operatorContext.getDriverContext()... here, this should be done through the localUserMemoryContext